subprojects/core/src/main/java/org/gradle/internal/progress/DefaultBuildOperationExecutor.java

```diff
@@ -16,133 +16,187 @@
+                LOGGER.debug("Build operation '{}' started", description.getDisplayName());
                 try {
-                    result = factory.transform(context);
+                    if (buildOperation instanceof RunnableBuildOperation) {
+                        ((RunnableBuildOperation) buildOperation).run(context);
+                    }
+                    if (buildOperation instanceof CallableBuildOperation) {
+                        result = Cast.uncheckedCast(((CallableBuildOperation) buildOperation).call(context));
+                    }
                 } finally {
                     if (progressLogger != null) {
                         progressLogger.completed();
                     }
                 }
                 if (parent != null && !parent.isRunning()) {
-                    throw new IllegalStateException(String.format("Parent operation (%s) completed before this operation (%s).", parent.operationDetails.getDisplayName(), operationDetails.getDisplayName()));
+                    throw new IllegalStateException(String.format("Parent operation (%s) completed before this operation (%s).",
+                        parent.getDescription().getDisplayName(), description.getDisplayName()));
                 }
             } catch (Throwable t) {
                 context.failed(t);
                 failure = t;
             }
 
             long endTime = timeProvider.getCurrentTime();
-            listener.finished(operation, new OperationResult(startTime, endTime, context.failure, context.result));
+            listener.finished((BuildOperationInternal) description, new OperationResult(currentOperation.getStartTime(), endTime, context.failure, context.result));
 
             if (failure != null) {
                 throw UncheckedException.throwAsUncheckedException(failure, true);
             }
 
-            LOGGER.debug("Build operation '{}' completed", operation.getDisplayName());
+            LOGGER.debug("Build operation '{}' completed", description.getDisplayName());
             return result;
         } finally {
-            if (parent instanceof UnmanagedThreadOperation) {
+            if (parent != null && parent.getDescription().getId() < 0) { //unmanaged thread
                 this.currentOperation.set(null);
                 currentOperation.setRunning(false);
-                UnmanagedThreadOperation unmanagedParent = (UnmanagedThreadOperation) parent;
                 try {
-                    listener.finished(unmanagedParent.internal, new OperationResult(unmanagedParent.startTime, timeProvider.getCurrentTime(), null, null));
+                    listener.finished((BuildOperationInternal) parent.getDescription(), new OperationResult(parent.getStartTime(), timeProvider.getCurrentTime(), null, null));
                 } finally {
-                    unmanagedParent.setRunning(false);
+                    currentOperation.setRunning(false);
```

Comment:
Argh. Good catch. We had the issue earlier with the unmanaged thread operations that needed to be used for multiple children in the case of parallel execution.

This was not quite right now. I moved the special handling of unmanaged thread out of `execute()` and put it around every call to `execute()` **and** `executeInParallel()`. I Added test coverage.